src/Microsoft.AspNet.Hosting/Internal/RequestServicesContainerMiddleware.cs

```diff
@@ -21,9 +23,38 @@ public RequestServicesContainerMiddleware(RequestDelegate next, IServiceProvider
 
         public async Task Invoke(HttpContext httpContext)
         {
-            using (var container = RequestServicesContainer.EnsureRequestServices(httpContext, _services))
+            // All done if we already have a request services
+            if (httpContext.RequestServices != null)
             {
-                await _next.Invoke(httpContext);
+                return;
+            }
+
+            var serviceProvider = httpContext.ApplicationServices ?? _services;
+
+            var appServiceProvider = serviceProvider.GetRequiredService<IServiceProvider>();
+            var appServiceScopeFactory = serviceProvider.GetRequiredService<IServiceScopeFactory>();
+
+            if (serviceProvider != appServiceProvider)
+            {
+                appServiceProvider = serviceProvider;
+                appServiceScopeFactory = serviceProvider.GetRequiredService<IServiceScopeFactory>();
```

Comment:
Does appServiceScopeFactory need to be resolved twice? Could you just resolve it once after setting appServiceProvider?